btree/btree.c

```diff
@@ -1743,16 +1796,56 @@ static int64_t node_get(struct node_op *op, struct td *tree,
 	op->no_node = nt->nt_opaque_get(addr);
 	if (op->no_node != NULL &&
 	    op->no_node->n_addr.as_core == addr->as_core) {
-		    if (op->no_node->n_delayed_free) {
-			    op->no_op.o_sm.sm_rc = EACCES;
-				return nxt_state;
+
+		m0_rwlock_write_lock(&op->no_node->n_lock);
+		if (op->no_node->n_delayed_free) {
+			op->no_op.o_sm.sm_rc = EACCES;
+			m0_rwlock_write_unlock(&op->no_node->n_lock);
+			return nxt_state;
 		}
-		op->no_node->n_ref++;
-	    } else {
+
+		in_lrulist = op->no_node->n_ref == 0;
+		node_refcnt_update(op->no_node, true);
+		if (in_lrulist) {
+			/**
+			 * The node descriptor is in LRU list. Remove from lru
+			 * list and add to trees active list
+			 */
+
+			m0_rwlock_write_lock(&lru_lock);
+			/* M0_ASSERT_EX(ndlist_tlist_contains(&btree_lru_nds,
+			 *                                    op->no_node)); */
+			ndlist_tlist_del(op->no_node);
+			m0_rwlock_write_unlock(&lru_lock);
+
+			tree_lock(tree, lock_acquired);
+			ndlist_tlist_add(&tree->t_active_nds, op->no_node);
+			tree_unlock(tree, lock_acquired);
+			/**
+			 * Update nd::n_tree  to point to tree descriptor as we
+			 * as we had set it to NULL in node_put(). For more
+			 * details Refer comment in node_put().
+			 */
+			op->no_node->n_tree = tree;
+		}
+		m0_rwlock_write_unlock(&op->no_node->n_lock);
+	} else {
 		/**
 		 * If node descriptor is not present allocate a new one
-		 * and assign to node. */
+		 * and assign to node.
+		 */
+		/**
+		 * TODO: Adding lru_lock to protect from two threads allocating
+		 * two node descriptors for the same node. Replace it with a
```

Comment:
comment should be modified; it is not necessary that there will be two threads only, you might want o  use more than one thread , more than one node descriptor than two thread and two node descriptor , respectively.

Reply:
Sure, will modify comment.